src/main/java/com/emc/pravega/perf/WriterWorker.java

```diff
@@ -55,33 +57,82 @@
+            CompletableFuture retFuture = null;
+            Random rand = new Random();
+
+            for (int i = 0; i < events; i++) {
+
+                // Construct event payload
+                String val = System.currentTimeMillis() + ", " + workerID + ", " + (int) (Math.random() * 200);
+                String payload = String.format("%-" + messageSize + "s", val);
+                String key;
+                if (isRandomKey) {
+                    key = Integer.toString(workerID + rand.nextInt());
+                } else {
+                    key = Integer.toString(workerID);
+                }
+
+                final Instant startTime = Instant.now();
+                retFuture = writeData(key, payload);
+                // event ingestion
+                retFuture = stats.recordTime(retFuture, startTime, payload.length());
             }
 
-            final Instant startTime = Instant.now();
-            retFuture = writeData(key, payload);
-            // event ingestion
-            retFuture = stats.recordTime(retFuture, startTime, payload.length());
+            flush();
+
+            //Wait for the last packet to get acked
+            retFuture.get();
+        }
+    }
+
+    private class eventspersecWriter implements performance {
+        private static final long NS_PER_MS = 1000000L;
+        private static final long NS_PER_SEC = 1000 * NS_PER_MS;
+        private static final long MIN_SLEEP_NS = 100000L;
+        private final long sleep_ns, sleep_ms, remain_ns;
+
+        eventspersecWriter() {
+            sleep_ns = NS_PER_SEC / events;
+            sleep_ms = sleep_ns / NS_PER_MS;
+            remain_ns = sleep_ns - sleep_ms * NS_PER_MS;
         }
 
-        flush();
+        public void benchmark() throws InterruptedException, ExecutionException {
+            CompletableFuture retFuture = null;
+            Random rand = new Random();
 
-        //Wait for the last packet to get acked
-        retFuture.get();
+            while (Duration.between(StartTime, Instant.now()).getSeconds() < secondsToRun) {
+                // Construct event payload
+                String val = System.currentTimeMillis() + ", " + workerID + ", " + (int) (Math.random() * 200);
+                String payload = String.format("%-" + messageSize + "s", val);
+                String key;
+                if (isRandomKey) {
+                    key = Integer.toString(workerID + rand.nextInt());
+                } else {
+                    key = Integer.toString(workerID);
+                }
 
-        return null;
+                final Instant beginTime = Instant.now();
+                retFuture = writeData(key, payload);
+                // event ingestion
+                retFuture = stats.recordTime(retFuture, beginTime, payload.length());
+
+                if (sleep_ns > MIN_SLEEP_NS) {
+                    Thread.sleep(sleep_ms, (int) remain_ns);
```

Comment:
Make sure you handle over-sleeping. The OS can't guarentee thread wake up time.

Reply:
hi Tom, for oversleeping, I have done the time adjustment for next iteration ; but now its part of throughput controller